chain/network/src/network_protocol/borsh_conv.rs

```diff
@@ -0,0 +1,159 @@
+            net::HandshakeFailureReason::GenesisMismatch(genesis_id) => {
+                mem::HandshakeFailureReason::GenesisMismatch(genesis_id.clone())
+            }
+            net::HandshakeFailureReason::InvalidTarget => {
+                mem::HandshakeFailureReason::InvalidTarget
+            }
+        })
+    }
+}
+
+impl TryFrom<&mem::HandshakeFailureReason> for net::HandshakeFailureReason {
+    type Error = anyhow::Error;
+    fn try_from(x: &mem::HandshakeFailureReason) -> anyhow::Result<Self> {
+        Ok(match x {
+            mem::HandshakeFailureReason::ProtocolVersionMismatch {
+                version,
+                oldest_supported_version,
+            } => net::HandshakeFailureReason::ProtocolVersionMismatch {
+                version: *version,
+                oldest_supported_version: *oldest_supported_version,
+            },
+            mem::HandshakeFailureReason::GenesisMismatch(genesis_id) => {
+                net::HandshakeFailureReason::GenesisMismatch(genesis_id.clone())
+            }
+            mem::HandshakeFailureReason::InvalidTarget => {
+                net::HandshakeFailureReason::InvalidTarget
+            }
+        })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&net::PeerMessage> for mem::PeerMessage {
+    type Error = anyhow::Error;
+    fn try_from(x: &net::PeerMessage) -> anyhow::Result<Self> {
+        Ok(match x.clone() {
+            net::PeerMessage::Handshake(h) => mem::PeerMessage::Handshake((&h).try_into()?),
+            net::PeerMessage::HandshakeFailure(pi, hfr) => {
+                mem::PeerMessage::HandshakeFailure(pi, (&hfr).try_into()?)
+            }
+            net::PeerMessage::LastEdge(e) => mem::PeerMessage::LastEdge(e),
+            net::PeerMessage::SyncRoutingTable(rtu) => mem::PeerMessage::SyncRoutingTable(rtu),
+            net::PeerMessage::RequestUpdateNonce(e) => mem::PeerMessage::RequestUpdateNonce(e),
+            net::PeerMessage::ResponseUpdateNonce(e) => mem::PeerMessage::ResponseUpdateNonce(e),
+            net::PeerMessage::PeersRequest => mem::PeerMessage::PeersRequest,
+            net::PeerMessage::PeersResponse(pis) => mem::PeerMessage::PeersResponse(pis),
+            net::PeerMessage::BlockHeadersRequest(bhs) => {
+                mem::PeerMessage::BlockHeadersRequest(bhs)
+            }
+            net::PeerMessage::BlockHeaders(bhs) => mem::PeerMessage::BlockHeaders(bhs),
+            net::PeerMessage::BlockRequest(bh) => mem::PeerMessage::BlockRequest(bh),
+            net::PeerMessage::Block(b) => mem::PeerMessage::Block(b),
+            net::PeerMessage::Transaction(t) => mem::PeerMessage::Transaction(t),
+            net::PeerMessage::Routed(r) => mem::PeerMessage::Routed(r),
+            net::PeerMessage::Disconnect => mem::PeerMessage::Disconnect,
+            net::PeerMessage::Challenge(c) => mem::PeerMessage::Challenge(c),
+            net::PeerMessage::_HandshakeV2 => bail!("HandshakeV2 is deprecated"),
```

Comment:
You should validate all the data that comes from outside of the machine that your process is running on (in fact, even data from other processes should be highly suspicious).
"unreachable!" would make sending a _HandshakeV2 to a node a "query of doom", so it is out of question.
Hence, we should treat _HandshakeV2 as a parsing error (where I consider conversion net::PeerMessage -> mem::PeerMessage a step of parsing). Here I'm encapsulating the net::PeerMessage as an implementation detail of the message parser, exactly to not introduce any changes to it - I want to get rid of it exactly because introducing changes to it may break the network protocol backward compatibility.

Reply:
Ack! (but consider marking conversions which could be infallible as such)